### DISCUSSION TITLE
Who is using PF4J?

### DISCUSSION THREAD

**Discussion by decebals** [pf4j/pf4j#620] [General]
This thread collects projects and products that use PF4J. The README carries a short list of the best known ones, this is the long one.

**Verified.** PF4J is visible in the source, in the build, or in the product's own output.

- [Netflix Spinnaker](https://github.com/spinnaker/spinnaker) - continuous delivery platform, its plugin SDK is built on PF4J
- Facebook Buck (Java version) - build system
- Huawei Cloud MRS - real-time data processing and scoring systems
- Eclipse Foundation ECSP - connected services platform for software defined vehicles
- Dremio, now part of SAP - data lakehouse, with its own plugin manager and class loaders on top of PF4J
- MuleSoft Anypoint Code Builder, Salesforce - the Mule DX Runtime loads its IDE component servers as PF4J plugins
- [Apache Tika](https://github.com/apache/tika) - plugin ZIPs built at build time, loaded in deployment mode
- [Apache BifroMQ](https://github.com/apache/bifromq) - distributed MQTT broker, plugins for auth, throttling, event collection and settings
- [Appsmith](https://github.com/appsmithorg/appsmith) - low-code platform, its datasource connectors are PF4J plugins
- [Nextflow](https://github.com/nextflow-io/nextflow) - workflow engine for scientific pipelines
- [ReportPortal](https://github.com/reportportal/reportportal) - test automation dashboard, uses pf4j-update for its plugin marketplace
- [Halo](https://github.com/halo-dev/halo) - content management
- [MeterSphere](https://github.com/metersphere/metersphere) - test management platform
- [nGrinder](https://github.com/naver/ngrinder), Naver - load testing

**Reported in this thread or found as dependents**, not checked in detail: [Lowcoder](https://github.com/lowcoder-org/lowcoder), [FSCrawler](https://github.com/dadoonet/fscrawler), [Turms](https://github.com/turms-im/turms), [PagePlug](https://github.com/cloudtogo/pageplug), [Gitblit](https://github.com/gitblit/gitblit), [Sparkler](https://github.com/USCDataScience/sparkler), [CFLint](https://github.com/cflint/CFLint), [Gluu Casa](https://github.com/GluuFederation/casa), [ServerPackCreator](https://github.com/Griefed/ServerPackCreator), [Sandpolis](https://github.com/Subterranean-Security/Sandpolis), [OpenKM](https://github.com/terwer/openkm), [open-wes](https://github.com/jingsewu/open-wes), [LegendaryBot](https://github.com/greatman/legendarybot).

Through pf4j-spring: [PeerBanHelper](https://github.com/PBH-BTN/PeerBanHelper), [yudao-cloud](https://github.com/YunaiV/yudao-cloud), [ruoyi-vue-pro](https://github.com/YunaiV/ruoyi-vue-pro). There is also a [Jmix plugin addon](https://github.com/Gamer08YT/jmix-plugin-addon) built on PF4J.

If you use PF4J, leave a comment with a link and I will add it.

**Comment by styurin** [comment]
Hey, I'm considering using it for buck (https://github.com/facebook/buck/).

**Comment by josiahhaswell** [comment]
Sunshower.io (https://sunshower.io) (need to update SSL certificate) is considering using it.

**Comment by greatman** [comment]
Using it here https://github.com/greatman/legendarybot

**Comment by myronlstewart** [comment]
I’m using it in a project called IngenoGears and using it in another project. I like it because it’s easier to implement than OSGI.

**Comment by justinnichols** [comment]
I didn't see this until now.  We're using version 1.0.0 right now in a plugin system I created to run multiple jobs independently scheduled (using quartz or using akka on Play).  We're looking to update to the latest PF4J soon.

**Comment by decebals** [comment]
https://github.com/gitblit/gitblit
https://github.com/USCDataScience/sparkler
https://github.com/naver/ngrinder
https://github.com/cflint/CFLint
https://github.com/apache/ignite

**Comment by jgomer2001** [comment]
We'll use it for the upcoming version of Gluu's [cred-manager](https://github.com/GluuFederation/cred-manager/) which has been rebranded to [Gluu Casa](https://github.com/GluuFederation/casa)

**Comment by pwysocki7** [comment]
@justinnichols link to your plugin system?

**Comment by HJacco** [comment]
I  will use it to refactor our pay system, every pay channel will be a plugin

**Comment by co5dt** [comment]
We are building a CTI SaaS platform and are in the kick-off phase of things. Playing around with it. Still having many questions and wish for a more extensive documentation. Looks promising though.

**Comment by cilki** [comment]
I recently added PF4J to [https://github.com/Subterranean-Security/Sandpolis](https://github.com/Subterranean-Security/Sandpolis). I'm not finished, but it's working extremely well so far. Thanks for this library :+1:.

**Comment by Griefed** [comment]
I've added pf4j to my passion project, [ServerPackCreator](https://github.com/Griefed/ServerPackCreator), a while ago. It's a tool which generates server packs (the multiplayer component) of Minecraft modpacks, so they can then be distributed over the webz.

pf4j is a godsend. Without it I would probably not have been able to provide the desired addon-functionality that SPC is getting now.

**Comment by terwer** [comment]
I started an opensource KM(Knowledge management) based plantform using it.I benefit a lot from it.
**It seperates big projects to small pices, but also can run along with** .This is very cool.

https://github.com/terwer/openkm

**Comment by guqing** [comment]
**Halo** has implemented a comprehensive plugin system using pf4j:
https://github.com/halo-dev/halo 
It has 31.1k+ stars on github
Thank you to pf4j for such an excellent project♥️

**Comment by Gamer08YT** [comment]
I use PF4J in several projects, my current main project is:

![Image](https://github.com/user-attachments/assets/f2c9f35e-49a3-49fb-942e-bdd05cce06c7)

Hostinger offers the possibility to manage VMs / containers as well as processes, domains and web spaces. It also offers a CMS and CRM system as well as some e-commerce functions.

[https://byte-storede.github.io/Hostinger-Docs/hostinger.html](https://byte-storede.github.io/Hostinger-Docs/hostinger.html)

My main project is based on the [Jmix ](https://github.com/jmix-framework/jmix)Plugin Addon, which is based on PF4J.

![Image](https://github.com/user-attachments/assets/a444e00e-6a76-475f-bf46-07ed3c1b996e)

[https://github.com/Gamer08YT/jmix-plugin-addon/](https://github.com/Gamer08YT/jmix-plugin-addon/)

This will appear in the Marketplace with [Release 2.5](https://github.com/jmix-framework/jmix/pull/4189).

**Comment by jingsewu** [comment]
I am leveraging PF4J to enhance my open-source warehouse execution system. Integrating PF4J into a full-scale business system (as opposed to a standalone tool) poses unique technical challenges. To improve extensibility, I am also developing a dedicated plugin store for seamless integration of modular components. The project link is: https://github.com/jingsewu/open-wes

**Comment by decebals** [comment]
https://github.com/apache/tika
